File: 1_matrix_vector_dot_product.py
def matrix_dot_vector(a: list[list[int or float]], b: list[int or float]) -> list[int or float]:
    if len(a[0]) != len(b):
        return -1
    
    result = []
    for row in a:
        total = 0
        for i in range(len(row)):
            total += row[i] * b[i]
        result.append(total)
    return result

print(matrix_dot_vector([[1, 2, 3], [2, 4, 5], [6, 8, 9]], [1, 2, 3]))

# explanation:
#  a = [[1, 2, 3],  b = [1, 2, 3]
#       [2, 4, 5], 
#       [6, 8, 9]]

# row 1: 1*1 + 2*2 + 3*3 = 1 + 4 + 9 = 14
# row 2: 2*1 + 4*2 + 5*3 = 2 + 8 + 15 = 25
# row 3: 6*1 + 8*2 + 9*3 = 6 + 16 + 27 = 49

#  result = [1*1 + 2*2 + 3*3, 
#            2*1 + 4*2 + 5*3, 
#            6*1 + 8*2 + 9*3]

#  result = [14, 25, 49]

File: 1_matrix_vector_dot_product.rs
fn matrix_dot_vector(a: &[Vec<f64>], b: &[f64]) -> Result<Vec<f64>, &'static str> {
  if a.is_empty() || a[0].len() != b.len() {
    return Err("Dimension mismatch")
  }
  
  let mut result = Vec::new();
  for row in a {
    let mut total = 0.0;
    for (i, &val) in row.iter().enumerate() {
      total += val * b[i];
    }
    result.push(total);
  }

  Ok(result)
}

// Example usage and test
fn main() {
  let matrix = vec![
      vec![1.0, 2.0, 3.0],
      vec![4.0, 5.0, 6.0],
      vec![7.0, 8.0, 9.0],
  ];
  let vector = vec![1.0, 2.0, 3.0];
  
  match matrix_dot_vector(&matrix, &vector) {
      Ok(result) => println!("Result: {:?}", result),
      Err(e) => println!("Error: {}", e),
  }
  
  // Test with mismatched dimensions
  let bad_vector = vec![1.0, 2.0];
  match matrix_dot_vector(&matrix, &bad_vector) {
      Ok(result) => println!("Result: {:?}", result),
      Err(e) => println!("Error: {}", e),
  }
}
